### PR TITLE
Use custom color field classes in all models

### LINKED ISSUE #3884: Make custom pydantic color validator
## 🧰 Task

Following discussion in https://github.com/napari/napari/pull/3879#issuecomment-1000922569 we should make a custom color [pydantic type with its own validator](https://pydantic-docs.helpmanual.io/usage/types/#classes-with-__get_validators__) based on

https://github.com/napari/napari/blob/43aa65bcde1ef36493ba51499edca4fad8d7fafe/napari/components/text_overlay.py#L37-L39

and then reuse it in all places where we currently have

https://github.com/napari/napari/blob/43aa65bcde1ef36493ba51499edca4fad8d7fafe/napari/components/text_overlay.py#L32

or https://github.com/napari/napari/blob/9152dbd7c328b12a9c3b19c9629f58906650465c/napari/layers/surface/wireframe.py#L24

This will mean we no longer need to use the union of types, but always have one nice type

### PR DESCRIPTION
**Pull request by andy-sweet**
# Description

This uses our custom type to consistently handle validating/coercing pydantic model fields that represent single and multi-color values by using the `ColorValue` and `ColorArray` classes made public in #4464. I moved those definitions to the newly created and more visible `napari.utils.color` module as these are used more now.

I also wrote detailed docstrings for `ColorValue` and `ColorArray` to describe what input is supported and pointed the fields that use these type to those docstrings for more details, to avoid copy/pasting docstrings and the errors associated with that.

I did not use these types in `ColorManager` because that would cause a small change in behavior and we are likely to deprecate/drop that class in the not too distant future.

## Type of change
- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

This breaks a few things in minor ways.

- Moved `ColorValue` and `ColorArray` from `napari.layers.utils.color_encoding` to `napari.utils.color`. This breaks `main`, but not the last public release.
- Renamed `ColorValue.validate_type` to `ColorValue.validate` (same for `ColorArray`) because that's shorter with no less meaning and consistent with [the pydantic docs](https://pydantic-docs.helpmanual.io/usage/types/#custom-data-types) and `StringValue`. This breaks `main`, but not the last public release.
- Changes `Normals` and `SurfaceWireframe` to store `ndarray` using these fields rather a `Union`. This was the original motivation from #3879, but I still want to highlight the change in behavior.

# References
Closes #3884

# How has this been tested?
- [ ] all existing tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/utils/colormaps/standardize_color.py
Comment (andy-sweet) [inline napari/utils/colormaps/standardize_color.py:98]: This is a leftover from #4666

File: napari/layers/surface/normals.py
Comment (brisvag) [inline napari/layers/surface/normals.py]: What does this todo refer to? The fact that it stores the `ColorValue` rather than the string? If so, I think that's fine. The downside is maybe that you can't do `ColorValue == 'red'`...

Comment (andy-sweet) [reply napari/layers/surface/normals.py]: Yeah, I left it there to highlight the change in behavior as most other changes here don't change anything.

But your comment gave me an idea. Why can't we have `__eq__` and get the behavior you described. So here's my attempt at that.

```python
import numpy as np
from pydantic import BaseModel
from napari.utils.colormaps.standardize_color import transform_color

class ColorValue(np.ndarray):
    def __new__(cls, value): 
        array = transform_color(value)[0]
        return array.view(cls)

    @classmethod
    def __get_validators__(cls):
        yield cls._validate

    @classmethod
    def _validate(cls, value):
        return cls(value)

    def __eq__(self, other):
        return np.array_equal(self, ColorValue(other))

class MyModel(BaseModel):
    color: ColorValue

c = ColorValue('red')
c # => ColorValue([1., 0., 0., 1.], dtype=float32)
c == 'red' # => True
c == [1, 0, 0] # => True
c == 'blue' # => False
# As it subclasses ndarray, it looks like one
isinstance(c, np.ndarray) # True
c.dtype # => dtype('float32')
c.shape # => (4,)

m = MyModel(color='red')
m.color # => ColorValue([1., 0., 0., 1.], dtype=float32)
m.color == 'red' # => True
```

Kinda nice, right? What do you think?

I'm pretty new to messing around with `__new__` and came up with this after reading [the numpy docs on subclassing `ndarray`](https://numpy.org/doc/stable/user/basics.subclassing.html), so there could definitely be some problem I don't see here or maybe some weird potential interactions with pydantic. I did find that yielding `cls.__new__` directly from `__get_validators__` caused a pydantic validation error with the message `Cannot change data-type for object array. (type=type_error)`, which is why I used `_validate` instead.

Comment (andy-sweet) [reply napari/layers/surface/normals.py]: Defining `__eq__` could also play nicely with `EventedModel` and I think would negate the need to add custom equality operators as we do now.

Comment (brisvag) [reply napari/layers/surface/normals.py]: This is awesome!

Comment (andy-sweet) [reply napari/layers/surface/normals.py]: Also curious to get @tlambert03 's opinion on the above before potentially including this here and making it a non-draft PR.

Comment (jni) [reply napari/layers/surface/normals.py]: I love this too! ❤️

File: napari/layers/utils/_tests/test_color_manager.py
Comment (jni) [inline napari/layers/utils/_tests/test_color_manager.py]: I don't love this idea! Where does the need come from to do anything at all with None or an empty list?

**Comment by jni** [comment]
So far I like what I see @andy-sweet!

File: napari/layers/utils/_tests/test_color_manager.py
Comment (andy-sweet) [reply napari/layers/utils/_tests/test_color_manager.py]: In napari it comes from `transform_color`. We specifically handle [`None`](https://github.com/napari/napari/blob/19775aaf949fa0f27be37b56d7422d60556cd749/napari/utils/colormaps/standardize_color.py#L354). We specifically handle the case where we get an [empty array-like](https://github.com/napari/napari/blob/19775aaf949fa0f27be37b56d7422d60556cd749/napari/utils/colormaps/standardize_color.py#L202), though we do issue a warning.

The `None` behavior is consistent with `vispy.color.Color`, but the empty array-like raises an error in vispy.

```python
In [3]: from vispy.color import Color

In [4]: c = Color(None)

In [5]: c
Out[5]: <Color: (0.0, 0.0, 0.0, 0.0)>

In [6]: c = Color([])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-2099ae45e5f3> in <module>
----> 1 c = Color([])

~/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/vispy/color/color_array.py in __init__(self, color, alpha, clip)
    387         if isinstance(color, (list, tuple)):
    388             color = np.array(color, np.float32)
--> 389         rgba = _user_to_rgba(color, clip=clip)
    390         if rgba.shape[0] != 1:
    391             raise ValueError('color must be of correct shape')

~/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/vispy/color/color_array.py in _user_to_rgba(color, expand, clip)
     54     color = np.atleast_2d(color).astype(np.float32)
     55     if color.shape[1] not in (3, 4):
---> 56         raise ValueError('color must have three or four elements')
     57     if expand and color.shape[1] == 3:  # only expand if requested
     58         color = np.concatenate((color, np.ones((color.shape[0], 1))),

ValueError: color must have three or four elements
```

For `ColorManager.current_color` we have special handling of `None` and `[]` when in `DIRECT` mode, which is what the test being changed here is handling. On `main` we get opaque black instead of white with warning after this PR. As we might remove `ColorManager` in the nearish future, maybe we shouldn't change its behavior in this PR.

But we should decide what we going forward. I'm happy with coercing `None` to `[0, 0, 0, 0]` like vispy as it kind of represents no color, but also happy to not coerce it. I do not think we should coerce an empty array-like though, which just feels like a classic `ValueError` that should raise.

File: napari/layers/surface/normals.py
Comment (tlambert03) [reply napari/layers/surface/normals.py]: yaaas

File: napari/layers/utils/_tests/test_color_manager.py
Comment (tlambert03) [reply napari/layers/utils/_tests/test_color_manager.py]: I'm fine with coercing falsey values to whatever the default for `None` is

**Review by tlambert03 — COMMENTED** [review]
this is excellent.  been wanting something going in this direction for a long time.

The only thing I think i still find myself wanting after this (and it could go in a later PR), is additional methods.  I'm fine with our colors all fundamentally looking like, or actually being, numpy arrays.  But I would love it if they had methods like `to_hex()`, `to_rgb()`, event `to_name()`, etc...  an `__array__` method would satisfy the need to use it _like_ a numpy array anywhere that's desireable.

I know this represents something of a fundamental change (since you'd need __new__ to return something other than a view), and I recognize that the general desire to make everything "just a numpy array" might preclude me ever having this :) ... but generally changing everything to `color: ColorValue` is absolutely a step in the right direction imo.

thanks @andy-sweet!

**Comment by andy-sweet** [comment]
This is ready for review now. See the description for some updated details.

I decided to avoid the `ndarray.view` approach for now because that adds a few more technical design decisions, whereas this PR represents an improvement over main without the need for those.

**Comment by andy-sweet** [comment]
For the record, here are the benchmarks showing that this approach is fast with vispy in cases where we have lots of colors.

```python
In [27]: color_ndarray = np.array(100000 * [[1, 0, 0, 0]], dtype=np.float32)

In [28]: color_list_of_lists = 100000 * [[1, 0, 0, 0]]

In [29]: color_list_of_tuples = 100000 * [(1, 0, 0, 0)]

In [30]: color_list_of_names = 100000 * ['red']

In [31]: %timeit vispy.color.ColorArray(color_ndarray)
443 µs ± 6.31 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [32]: %timeit vispy.color.ColorArray(color_list_of_lists)
44.5 ms ± 1.98 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [33]: %timeit vispy.color.ColorArray(color_list_of_tuples)
44.6 ms ± 1.52 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [34]: %timeit vispy.color.ColorArray(color_list_of_names)
1.5 s ± 45.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

We can see that relying on vispy to convert a sequence of names is probably not viable and using an `np.ndarray` is quite a bit faster than anything else.

**Comment by andy-sweet** [comment]
> The only thing I think i still find myself wanting after this (and it could go in a later PR), is additional methods. I'm fine with our colors all fundamentally looking like, or actually being, numpy arrays. But I would love it if they had methods like `to_hex()`, `to_rgb()`, event `to_name()`, etc... an `__array__` method would satisfy the need to use it _like_ a numpy array anywhere that's desireable.

I think this is all achievable with the [`ndarray.view`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.view.html) approach as it's using the `type` parameter (and not the `dtype` parameter) to return a class of type `ColorValue(ndarray)` rather than just `ndarray`. Though, methods like `to_hex` and `to_rgb` might be lossy (i.e. we return the value closest to the floating point representation), and `to_name` might return `None` if `ColorValue` was not created with a color name. If we wanted something more like pydantic's `Color`, then I think we'd also have to define how we want to handle multiple colors efficiently. I think it wouldn't be too hard, but I left it as out of scope for this PR.

**Comment by tlambert03** [comment]
> Though, methods like to_hex and to_rgb might be lossy (i.e. we return the value closest to the floating point representation), and to_name might return None if ColorValue was not created with a color name

yeah, both fine with me.   name is expected to be `None` if it's not a recognized color.  and the loss from floating point to hex seems like it is likely to be imperceptible.

> we'd also have to define how we want to handle multiple colors efficiently.

wait, wouldn't that be `ColorArray`?

**Comment by andy-sweet** [comment]
> wait, wouldn't that be `ColorArray`?

In this PR, where I haven't done the fancy `ndarray.view` thing yet, it is. The relationship between `ColorValue` and `ColorArray` is pretty simple as they both have type `ndarray`, dtype of `float32`, and shapes of `(4,)` and `(-1, 4)` respectively.

If we define a fancier `ColorValue` that implements `__eq__`, `to_hex`, `to_name`, etc. then that introduces slightly more inconsistency between the two classes. It also asks the question if `ColorArray` should have way to return an instance of that class (e.g. `get_element(index) -> ColorValue`) or if people should just be OK with using something like `ColorValue(colors[0])`. 

Kermit decided he had to urgently let me know about a fly he was hunting at 4am this morning, so my brain is just about of juice and this seems like a somewhat tricky design problem that I'd rather not think about right now. But happy to be convinced and may think differently with fresh eyes in the morning.

**Comment by andy-sweet** [comment]
> Kermit decided he had to urgently let me know about a fly he was hunting at 4am this morning, so my brain is just about of juice and this seems like a somewhat tricky design problem that I'd rather not think about right now. But happy to be convinced and may think differently with fresh eyes in the morning.

In the light of day, I still think this PR represents an improvement over main with more consistency of the currently accepted approach of coercing colors to numpy arrays, plus more documentation (in the form of docstrings) about the input color values that can be coerced to those.

Some of the other color design ideas are reasonable to consider, but I think they're future work and shouldn't block this PR.

**Review by sofroniewn — APPROVED** [review]
I like this, I prefer the new ColorValue and validation. Getting rid of all the `Array[float, (-1, 4)]` etc is good

**Review by jni — APPROVED** [review]

**Review by brisvag — APPROVED** [review]
Looks great, and we got several positive comments and reviews. I think we can merge!

**Pull request merged**
